grpc/src/main/java/com/linecorp/armeria/internal/common/grpc/GrpcStatus.java

```diff
@@ -81,7 +82,7 @@ public static Status fromThrowable(Throwable t) {
         if (s.getCode() != Code.UNKNOWN) {
             return s;
         }
-        if (t instanceof ClosedStreamException) {
+        if (t instanceof ClosedStreamException && !(t instanceof ClosedSessionException)) {
```

Comment:
Can we add a helper somewhere that just returns this boolean condition with documentation? It's not clear to me what the semantic meaning of this is. Or perhaps it's an overridden method of `ClosedStreamException`.

Though one thought is given the complexity of this sort of check, is it actually a good idea to have the inheritance relationship after all?

Reply:
> Can we add a helper somewhere that just returns this boolean condition with documentation?

That's good. Will add it to `Exceptions`.

> Though one thought is given the complexity of this sort of check, is it actually a good idea to have the inheritance relationship after all?

I know both approaches have cons. 
- Cons of inheritance: Devs should pay attention to filter an exception caused by a closed stream not closed session. 
- Cons of isolation: When a request fails to write a stream, users should catch both `ClosedStreamException` and `ClosedSessionException`. Because it could be failed by a closed session.

Reply:
I think it's common for methods to throw different types of exceptions and Java even has special syntax to catch `ClosedStreamException | ClosedSessionException`. 

On the flip side, filtering subclasses when working with anything, even non-exceptions is quite unusual. Inheritance should express "is a", so I don't find it to be idiomatic in Java to want to catch `ClosedStreamException` but not want it's subclasses, which are supposed to be semantically the same.

Reply:
> filtering subclasses when working with anything, even non-exceptions is quite unusual.

Yes, I understand what you are concerned about and greatly appreciate your opinion. 🙏 Let's discuss this with other folks at the beginning of next week. 😀

Reply:
My guess is that most users wouldn't need to distinguish between closed streams and closed sessions, while Armeria devs might in some cases, so it's probably better for users' standpoint?

Reply:
question: Don't we need to return `Status.CANCELLED` for `ClosedSessionException`?

Reply:
That's an interesting point. I've tried to keep the original condition. 
Any thoughts on this @anuraaga ?

Reply:
The test with reverting this, `GrpcServiceServerTest.clientSocketClosedBeforeHalfClose` is failed.
https://github.com/line/armeria/blob/46eff82ddbf52af751fb31debc2273638ae6a5f9/grpc/src/test/java/com/linecorp/armeria/server/grpc/GrpcServiceServerTest.java#L807-L808

The assert of the test expects `UNKNOWN` when the connection has been closed.

Reply:
Upstream only uses CANCELLED with `RST_STREAM`. Actually it is a bit more complex, looking at the http error code too which I guess we aren't but don't think we can catch connection failure here.

It's hard to know what a user would do but I guess they might implement a client type and run into this, or otherwise will generally just catch `Throwable` anyways.

Anyways I guess having this sort of complexity is ok, but it looks like wrong Java to me. It makes me question `ClosedStreamException` a little - this was meant to only map `RST_STREAM` but is that still the case?

Also what if stream extends session instead other way around? There is generic concept of session, whild HTTP/2 specializes that to the concept of a stream? Semantically it seems just a little better.

Reply:
Hi, all. I'm a fan and a user of Armeria and I'm interested in this topic. Would you mind I join the discussion? :-) What's about to create a new base class, let's say `BaseClosedException`, and make both `ClosedSessionException` and `ClosedStreamException` to extend it?

Reply:
Hi @alex-lx, thanks a *lot* for joining in! Sure, that also could be an option, but I guess we need to name it properly. Another issue is that `ClosedStreamException` is in `com.linecorp.armeria.common.stream` while `ClosedSessionException` is in `com.linecorp.armeria.common`. Which package should the parent exception go to?

Reply:
> Hi, all. I'm a fan and a user of Armeria and I'm interested in this topic.

♥️ Welcome @alex-lx

> What's about to create a new base class, let's say `BaseClosedException`

Yes. That is another option that an exception includes `ClosedSessionException` and `ClosedStreamException`.

Let's say a user make their own `RetryStrategy`:
```
RetryStrategy.onStatus((status, cause) -> {
    // retry if a request fails to write a stream
    if (cause instanceof ClosedStreamException) {
        return backoff;
    }
    return null;
});
```

If we make `ClosedSessionException` and `ClosedStreamException` extend `BaseClosedException`, a user needs to know the exceptions' hierarchy. I'm not sure this is best for Armeria users. 🤔

Reply:
If we want to keep the class hierarchy and need to return `UNKNOWN` for `ClosedSessionException` and `CANCELLED` for `ClosedStreamException`, don't we need to always check if the exception is the subclass or not, first? So that we can get rid of this complex condition?
```suggestion
if (t instanceof ClosedSessionException | t instanceof ClosedChannelException) { // Remove line 88 as well.
    return s;
}
if (t instanceof ClosedStreamException) {
    return Status.CANCELLED;
}
```

Reply:
> If we make ClosedSessionException and ClosedStreamException extend BaseClosedException, a user needs to know the exceptions' hierarchy. I'm not sure this is the best choice for Armeria users.

I think it's fine since we didn't release 1.0 just yet :-)

Reply:
I think one reasonable naming would for a base class would be `ClosedByPeerException` - both are the message being closed by the peer. One type of closing is a connection close, another is a HTTP/2 `RST_STREAM`.

Reply:
> let's say BaseClosedException, and make both ClosedSessionException and ClosedStreamException to extend it?

The name of `BaseClosedException` is a little vague.
How about reviving `ClosedPublisherException`(#2468) and making `ClosedSessionException and `ClosedStreamException` extend it?

Reply:
> `ClosedByPeerException`

I think we use the exception even if it's not closed by the peer. For example:
https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/client/HttpRequestSubscriber.java#L151

Reply:
Perhaps, we can define another exception for other cases and use `ClosedStreamException` and `ClosedSessionException` only when it's closed by the peer?

Reply:
> Other cases mean that the client and server want to close the stream and session.

I guess it may not always be easy to determine who closed stream/session and there's nothing much a user can do with such distinction.

Reply:
After thinking a little bit more, I'm leaning toward the original idea of leaving this PR as it is. Here's why:

- A user doesn't really care about why the stream was aborted but only about whether the stream was aborted or not.
  - If a user wants to know why, then he or she can catch `ClosedSessionException` first, e.g.
    ```java
    try { ... }
    catch (ClosedSessionException e) { ... }
    catch (ClosedStreamException e) { ... }
    // or
    if (e instanceof ClosedSessionException) { ... }
    else if (e instanceof ClosedStreamException) { ... }
    ```
    .. and this isn't that complex. It's like catching `FileNotFoundException` before `IOException`.
- I'm not sure if a common base exception helps the situation. It helps users catch the 'stream closed but not disconnected' case easily, but again, I don't think users are interested in this in most cases. If interested, an extra `instanceof` or `catch()` does the job.

Reply:
Ok SGTM